text/0000-no-entry.md

```diff
@@ -0,0 +1,149 @@
+will cause Rust to omit the entry-point symbol when compiling and linking the
+binary. When linking, Rust will pass any necessary target-specific arguments to
+the toolchain to omit the target's entry-point code, such as `-nostartfiles`.
+
+Binaries built with this attribute set will typically not link unless the user
+supplies their own entry-point symbol.
+
+Libraries cannot use this attribute; attempting to do so will produce a
+compile-time error stating that `no_entry` only applies to binaries.
+
+A binary built with this attribute set could supply the entry-point symbol
+itself, or have the entry-point symbol supplied by a library crate or by a
+non-Rust library.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Until stabilized, the `no_entry` attribute will require a feature-gate named
+`no_entry`.
+
+The `no_entry` attribute may be specified as part of a `cfg_attr` declaration,
+to omit the entry point in some configurations but not others.
+
+Specifying the `no_entry` attribute on anything other than the top-level module
+of a binary crate will produce an error (`warning: crate-level attribute should
+be in the root module`).
+
+Declaring an entry-point symbol will typically require a declaration with
+`#[no_mangle] pub extern "C"`.
+
+To implement `no_entry`, the compiler can have a default linker argument based
+on the type of linker in use, such as `-nostartfiles` for GCC or clang/LLD.
+Specific targets which require more specific options can provide those options
+via the target configuration structures.
+
+Some targets may be able to support invoking the C library without running the
+startup code provided by the C library; other targets may not support this, or
+have limitations on such support. Targets that do not support this should
+document the target-specific limitations of `no_entry` for their target, may
+require `no_std` in addition to `no_entry`, or may not support `no_entry` at
+all.
+
+On targets that already don't provide an entry point, the compiler will
+silently accept `no_entry`. On targets that cannot support `no_entry`, the
+compiler will emit a compile-time error.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Rust has several existing attributes related to program startup, and this would
+require further careful documentation to distinguish them. However, no existing
+attribute serves the function that `no_entry` does.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Today, people do this either by linking their binary manually, or by specifying
+non-portable linker arguments in `.cargo/config`. Introducing the `no_entry`
+attribute substantially simplifies creating such programs, and avoids
+target-specific build steps.
+
+The name `no_entry` was selected specifically to minimize conflict and
+confusion with other Rust features such as `start`, `no_main`, and `main`, none
```

Comment:
To better understand the cost of this proposal, I would like to see an appendix of all the current knobs (stable and unstable, with a note of which is which) regarding startup / entrypoint behavior. Ideally appendix would include a "flow chart" description.

Reply:
> To better understand the cost of this proposal, I would like to see an appendix of all the current knobs (stable and unstable, with a note of which is which) regarding startup / entrypoint behavior

I think it would be reasonable to have all the current knobs documented in a central place somewhere (e.g. the rustc user guide). I think it would be worth it to open an issue in rust-lang/rust about this. Once that happens, this and future RFCs can just reference to those docs and mention the additions.

Reply:
Sure, that also works.